### DISCUSSION TITLE
creating MUI lock screen

### DISCUSSION THREAD

**Discussion by Claponi** [olikraus/u8g2#2515] [Q&A] [answered]
Hi all,
I have a question regarding the creation of a lock screen.

My ideas was to create a copy of a form but with a fixed cursor. Let's call the copy form_2, and the original form_1.
form_1 is full of fields that can be changed... and when I click on a goto_button I go to form_2. There, I would like to only be able to click on the "same button" that send me back to form_1. That way, inside form_2, I still can see values entered in form_1 but can't modify them.

I'm sure there can be more complex lock screen, I was thinking this one was an easy to make. 

Sadly I'm stuck, cause I don't know how to fix the cursor position and where in the code I should do that (callback ?). Maybe something like :
```
uint8_t muif_LockBTN (mui_t *ui, uint8_t msg) {  
  uint8_t return_value = 0;
  switch (msg) {
    case MUIF_MSG_CURSOR_SELECT:
      return_value = mui_u8g2_btn_goto_wm_fi(ui, msg);
      if (ui-> is mud == 0){
      // get cursor position
      // block next field and previous field
     }
      break;
    default:
      return_value = mui_u8g2_btn_goto_wm_fi(ui, msg);
  }
  return return_value;
}
```

If you can help...

Thank you

**Comment by olikraus** [comment]
To avoid cursor movements in form_2, you could just stop sending the cursor movement commands. Lets consider this code: https://github.com/olikraus/u8g2/blob/eba8b1a0190ef4d9b3b3fd0ec907d9c039b31e9e/sys/arduino/u8g2_page_buffer/MUIInput3BtnBounce2/MUIInput3BtnBounce2.ino#L540-L554

You could check for form_2 (let me assume it has the number 2) and suppress any cursor movements. The modified code could look like this:

``` C++
void handle_events(void) {
   int form = mui.getCurrentFormId();;
  if ( selectBtn.pressed() ) {
    mui.sendSelect();
    is_redraw = 1;
  } 
  else if ( nextBtn.pressed() && from != 2) {
    mui.nextField();
    is_redraw = 1;
  }
  else if ( prevBtn.pressed()  && from != 2 ) {
    mui.prevField();
    is_redraw = 1;
  }    
}
```

**Reply by Claponi** [reply]
Thank you, simple and smart !

**Comment by Claponi** [comment]
Hi, I tried it today,
when accessing the new form the cursor go to the "initial cursor position", but that is not where is my button. 
I have to find a way to force the cursor to stay at the same position, this time maybe with a callback.

**Reply by olikraus** [reply, marked as answer]
I think you need to get the current position via `mui_GetCurrentCursorFocusPosition(mui.GetMUI())` and then later restore the cursor position with `gotoForm(uint8_t form_id, uint8_t initial_cursor_position)`.

In other words, you might need to create your own MUIF here, because the existing goto muif will always put the cursor into the first position.

Let me extend the code from here: https://github.com/olikraus/u8g2/blob/eba8b1a0190ef4d9b3b3fd0ec907d9c039b31e9e/csrc/mui_u8g2.c#L469-L497

``` C++
uint8_t mui_u8g2_btn_my_goto_wm_fi(mui_t *ui, uint8_t msg)
{
  uint8_t pos;
  switch(msg)
  {
    case MUIF_MSG_CURSOR_SELECT:
    case MUIF_MSG_VALUE_INCREMENT:
    case MUIF_MSG_VALUE_DECREMENT:
      pos = mui_GetCurrentCursorFocusPosition(ui);
      return mui_GotoForm(ui, ui->arg, pos);
    default:
      return mui_u8g2_btn_goto_wm_fi(ui, msg);
  }
  return 0;
}
```
... not tested...

I am just guessing here how you jump to the screen lock screen. If you have a timeout or so, it might be even simpler:

``` C++
      pos = mui_GetCurrentCursorFocusPosition(mui.GetMUI());
      mui.gotoForm(2, pos);
```

**Reply by Claponi** [reply]
thank you very much for all your help !!
